src/kernels/MIOpenNLLLoss.cpp

```diff
@@ -0,0 +1,441 @@
+    size_t Iidx = TV4D_IDX(input_tv, n[0], t, n[1], n[2]);
+    size_t Widx = TV1D_IDX(weight_tv, t);
+
+    if(t < 0 || t == ignore_index || t >= C)
+    {
+        output[Oidx] = static_cast<TO>(0);
+        return;
+    }
+
+    FLOAT_ACCUM w = weight != nullptr ? CVT_FLOAT2ACCUM(weight[Widx]) : CVT_FP32_2ACCUM(1.0f);
+
+    FLOAT_ACCUM input_value = CVT_FLOAT2ACCUM(input[Iidx]);
+
+    FLOAT_ACCUM val = CVT_FP32_2ACCUM(-1.0f) * w * input_value;
+    output[Oidx]    = CVT_ACCUM2FLOAT(val);
+}
+
+extern "C" __global__ void NLLLossUnreducedForward4d(const INPUT_TYPE* __restrict__ input,
+                                                     const int32_t* __restrict__ target,
+                                                     const INPUT_TYPE* weight,
+                                                     OUTPUT_TYPE* __restrict__ output,
+                                                     int32_t ignore_index,
+                                                     tensor_view_4d_t input_tv,
+                                                     tensor_view_3d_t target_tv,
+                                                     tensor_view_1d_t weight_tv,
+                                                     tensor_view_3d_t output_tv)
+{
+    nlllossUnreducedForward4d<INPUT_TYPE, OUTPUT_TYPE>(
+        input, target, weight, output, ignore_index, input_tv, target_tv, weight_tv, output_tv);
+}
+
+__device__ FLOAT_ACCUM warpReduceSum(FLOAT_ACCUM val)
+{
+    if(warpSize >= 64)
+        val += __shfl_down(val, 32);
+    if(warpSize >= 32)
+        val += __shfl_down(val, 16);
+    if(warpSize >= 16)
+        val += __shfl_down(val, 8);
+    if(warpSize >= 8)
+        val += __shfl_down(val, 4);
+    if(warpSize >= 4)
+        val += __shfl_down(val, 2);
+    if(warpSize >= 2)
+        val += __shfl_down(val, 1);
+    return val;
+}
+
+__device__ FLOAT_ACCUM blockReduceSum(FLOAT_ACCUM val)
+{
+    static __shared__ FLOAT_ACCUM shared[REDUCE_SIZE / warpSize];
+    auto lane = threadIdx.x % warpSize;
+    auto wid  = threadIdx.x / warpSize;
+
+    val = warpReduceSum(val);
+
+    if(lane == 0)
+        shared[wid] = val;
+    __syncthreads();
+
+    val = threadIdx.x < REDUCE_SIZE / warpSize ? shared[lane] : CVT_FP32_2ACCUM(0.0f);
+    if(wid == 0)
+        val = warpReduceSum(val);
+
+    return val;
+}
+
+template <typename DTYPE>
+__device__ void lossSum(const DTYPE* input, DTYPE* output, size_t N)
+{
+    auto gid = blockIdx.x * blockDim.x + threadIdx.x;
+
+    FLOAT_ACCUM val = gid < N ? CVT_FLOAT2ACCUM(input[gid]) : CVT_FP32_2ACCUM(0.0f);
+    val             = blockReduceSum(val);
+
+    if(threadIdx.x == 0)
+        output[blockIdx.x] = CVT_ACCUM2FLOAT(val);
+}
+
+extern "C" __global__ void
+LossSum(const D_TYPE* __restrict__ input, D_TYPE* __restrict__ output, size_t N)
+{
+    lossSum<D_TYPE>(input, output, N);
+}
+
+template <typename TI, typename TO>
+__device__ void nlllossReduceForward4d(TI* __restrict__ input,
```

Comment:
I think that modnn does not have the NLLLossReduceForward4d kernel. Did you create it yourself?